lib/mongoid/touchable.rb

```diff
@@ -22,46 +22,68 @@ module InstanceMethods
-        write_attribute(field, current) if field
-
-        # If the document being touched is embedded, touch its parents
-        # all the way through the composition hierarchy to the root object,
-        # because when an embedded document is changed the write is actually
-        # performed by the composition root. See MONGOID-3468.
-        if _parent
-          # This will persist updated_at on this document as well as parents.
-          # TODO support passing the field name to the parent's touch method;
-          # I believe it should be read out of
-          # _association.inverse_association.options but inverse_association
-          # seems to not always/ever be set here. See MONGOID-5014.
-          _parent.touch
-
-          if field
-            # If we are told to also touch a field, perform a separate write
-            # for that field. See MONGOID-5136.
-            # In theory we should combine the writes, which would require
-            # passing the fields to be updated to the parents - MONGOID-5142.
-            sets = set_field_atomic_updates(field)
-            selector = atomic_selector
-            _root.collection.find(selector).update_one(positionally(selector, sets), session: _session)
-          end
-        else
-          # If the current document is not embedded, it is composition root
-          # and we need to persist the write here.
-          touches = touch_atomic_updates(field)
-          unless touches["$set"].blank?
-            selector = atomic_selector
-            _root.collection.find(selector).update_one(positionally(selector, touches), session: _session)
-          end
-        end
+        write_attribute(:updated_at, now) if respond_to?("updated_at=")
+        write_attribute(field, now) if field
+
+        touch_parent = _parent && _association&.inverse_association&.touchable?
+
+        touches = __extract_touches_from_atomic_sets(field) || {}
+        touches.merge!(_parent.__gather_touch_updates(now) || {}) if touch_parent
+        touches
+      end
 
-        # Callbacks are invoked on the composition root first and on the
-        # leaf-most embedded document last.
+      # Recursively runs :touch callbacks for the document and its parents,
+      # beginning with the root document and cascading through each successive
+      # child document.
+      #
+      # @api private
+      def __run_touch_callbacks_from_root
+        _parent.__run_touch_callbacks_from_root if _parent
         run_callbacks(:touch)
-        true
+      end
+
+      private
+
+      # Extract and remove the atomic updates for the touch operation(s)
+      # from the currently enqueued atomic $set operations.
+      #
+      # @param [ Symbol ] field The optional field.
+      #
+      # @return [ Hash ] The field-value pairs to update atomically.
+      #
+      # @api private
+      def __extract_touches_from_atomic_sets(field = nil)
+        updates = atomic_updates['$set']
```

Comment:
Hmm... I'm keeping this consistent with the old `set_field_atomic_updates` method. Calling `atomic_updates` will build the updates, and I think `$set` will then include both `atomic_sets` and `delayed_atomic_sets`. If you're 100% confident there won't be knock-on effects, I'm fine to change, otherwise I'd say do it in a follow-up PR.